File: packages/backend/src/services/crypto.ts
import sodium from 'libsodium-wrappers';

/**
 * Cryptography service for end-to-end encryption
 * Uses libsodium (NaCl) for secure encryption
 */
export class CryptoService {
  private ready: boolean = false;

  constructor() {
    this.initialize();
  }

  async initialize(): Promise<void> {
    await sodium.ready;
    this.ready = true;
  }

  async ensureReady(): Promise<void> {
    if (!this.ready) {
      await this.initialize();
    }
  }

  /**
   * Generate a new keypair for user
   * @returns {Object} Public and private key pair
   */
  async generateKeyPair(): Promise<{ publicKey: string; privateKey: string }> {
    await this.ensureReady();
    const keyPair = sodium.crypto_box_keypair();
    return {
      publicKey: sodium.to_base64(keyPair.publicKey),
      privateKey: sodium.to_base64(keyPair.privateKey),
    };
  }

  /**
   * Encrypt a message for a recipient
   * @param message - Plain text message
   * @param recipientPublicKey - Recipient's public key (base64)
   * @param senderPrivateKey - Sender's private key (base64)
   * @returns Encrypted message (base64)
   */
  async encryptMessage(
    message: string,
    recipientPublicKey: string,
    senderPrivateKey: string
  ): Promise<string> {
    await this.ensureReady();

    const messageBytes = sodium.from_string(message);
    const recipientPubKey = sodium.from_base64(recipientPublicKey);
    const senderPrivKey = sodium.from_base64(senderPrivateKey);
    const nonce = sodium.randombytes_buf(sodium.crypto_box_NONCEBYTES);

    const ciphertext = sodium.crypto_box_easy(
      messageBytes,
      nonce,
      recipientPubKey,
      senderPrivKey
    );

    // Combine nonce and ciphertext
    const combined = new Uint8Array(nonce.length + ciphertext.length);
    combined.set(nonce);
    combined.set(ciphertext, nonce.length);

    return sodium.to_base64(combined);
  }

  /**
   * Decrypt a message
   * @param encryptedMessage - Encrypted message (base64)
   * @param senderPublicKey - Sender's public key (base64)
   * @param recipientPrivateKey - Recipient's private key (base64)
   * @returns Decrypted plain text message
   */
  async decryptMessage(
    encryptedMessage: string,
    senderPublicKey: string,
    recipientPrivateKey: string
  ): Promise<string> {
    await this.ensureReady();

    const combined = sodium.from_base64(encryptedMessage);
    const nonce = combined.slice(0, sodium.crypto_box_NONCEBYTES);
    const ciphertext = combined.slice(sodium.crypto_box_NONCEBYTES);

    const senderPubKey = sodium.from_base64(senderPublicKey);
    const recipientPrivKey = sodium.from_base64(recipientPrivateKey);

    const decrypted = sodium.crypto_box_open_easy(
      ciphertext,
      nonce,
      senderPubKey,
      recipientPrivKey
    );

    return sodium.to_string(decrypted);
  }

  /**
   * Encrypt message for a room (symmetric encryption)
   * @param message - Plain text message
   * @param roomKey - Room's shared key (base64)
   * @returns Encrypted message (base64)
   */
  async encryptRoomMessage(message: string, roomKey: string): Promise<string> {
    await this.ensureReady();

    const messageBytes = sodium.from_string(message);
    const key = sodium.from_base64(roomKey);
    const nonce = sodium.randombytes_buf(sodium.crypto_secretbox_NONCEBYTES);

    const ciphertext = sodium.crypto_secretbox_easy(messageBytes, nonce, key);

    // Combine nonce and ciphertext
    const combined = new Uint8Array(nonce.length + ciphertext.length);
    combined.set(nonce);
    combined.set(ciphertext, nonce.length);

    return sodium.to_base64(combined);
  }

  /**
   * Decrypt room message (symmetric decryption)
   * @param encryptedMessage - Encrypted message (base64)
   * @param roomKey - Room's shared key (base64)
   * @returns Decrypted plain text message
   */
  async decryptRoomMessage(
    encryptedMessage: string,
    roomKey: string
  ): Promise<string> {
    await this.ensureReady();

    const combined = sodium.from_base64(encryptedMessage);
    const nonce = combined.slice(0, sodium.crypto_secretbox_NONCEBYTES);
    const ciphertext = combined.slice(sodium.crypto_secretbox_NONCEBYTES);
    const key = sodium.from_base64(roomKey);

    const decrypted = sodium.crypto_secretbox_open_easy(ciphertext, nonce, key);

    return sodium.to_string(decrypted);
  }

  /**
   * Generate a random room key
   * @returns Room key (base64)
   */
  async generateRoomKey(): Promise<string> {
    await this.ensureReady();
    const key = sodium.randombytes_buf(sodium.crypto_secretbox_KEYBYTES);
    return sodium.to_base64(key);
  }

  /**
   * Hash data (for verification)
   * @param data - Data to hash
   * @returns Hash (base64)
   */
  async hash(data: string): Promise<string> {
    await this.ensureReady();
    const dataBytes = sodium.from_string(data);
    const hash = sodium.crypto_generichash(32, dataBytes);
    return sodium.to_base64(hash);
  }

  /**
   * Generate random bytes
   * @param length - Number of bytes
   * @returns Random bytes (base64)
   */
  async randomBytes(length: number): Promise<string> {
    await this.ensureReady();
    const bytes = sodium.randombytes_buf(length);
    return sodium.to_base64(bytes);
  }
}

export const cryptoService = new CryptoService();

File: rust-backend/src/services/crypto.rs
use crate::error::{AppError, Result};
use sodiumoxide::crypto::{box_, secretbox};
use sodiumoxide::randombytes;

pub struct CryptoService;

impl CryptoService {
    pub fn new() -> Self {
        sodiumoxide::init().expect("Failed to initialize sodiumoxide");
        Self
    }

    /// Generate keypair for user (asymmetric)
    pub fn generate_keypair(&self) -> Result<(String, String)> {
        let (public_key, secret_key) = box_::gen_keypair();

        Ok((
            base64::encode(public_key.as_ref()),
            base64::encode(secret_key.as_ref()),
        ))
    }

    /// Encrypt message for recipient (asymmetric)
    pub fn encrypt_message(
        &self,
        message: &str,
        recipient_public_key: &str,
        sender_private_key: &str,
    ) -> Result<String> {
        let recipient_pk = box_::PublicKey::from_slice(
            &base64::decode(recipient_public_key)
                .map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid public key".to_string()))?;

        let sender_sk = box_::SecretKey::from_slice(
            &base64::decode(sender_private_key).map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid private key".to_string()))?;

        let nonce = box_::gen_nonce();
        let ciphertext = box_::seal(message.as_bytes(), &nonce, &recipient_pk, &sender_sk);

        let mut combined = nonce.as_ref().to_vec();
        combined.extend_from_slice(&ciphertext);

        Ok(base64::encode(&combined))
    }

    /// Decrypt message (asymmetric)
    pub fn decrypt_message(
        &self,
        encrypted_message: &str,
        sender_public_key: &str,
        recipient_private_key: &str,
    ) -> Result<String> {
        let combined =
            base64::decode(encrypted_message).map_err(|e| AppError::Encryption(e.to_string()))?;

        if combined.len() < box_::NONCEBYTES {
            return Err(AppError::Encryption(
                "Invalid encrypted message".to_string(),
            ));
        }

        let nonce = box_::Nonce::from_slice(&combined[..box_::NONCEBYTES])
            .ok_or_else(|| AppError::Encryption("Invalid nonce".to_string()))?;

        let ciphertext = &combined[box_::NONCEBYTES..];

        let sender_pk = box_::PublicKey::from_slice(
            &base64::decode(sender_public_key).map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid public key".to_string()))?;

        let recipient_sk = box_::SecretKey::from_slice(
            &base64::decode(recipient_private_key)
                .map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid private key".to_string()))?;

        let decrypted = box_::open(ciphertext, &nonce, &sender_pk, &recipient_sk)
            .map_err(|_| AppError::Encryption("Decryption failed".to_string()))?;

        String::from_utf8(decrypted).map_err(|e| AppError::Encryption(e.to_string()))
    }

    /// Encrypt room message (symmetric)
    pub fn encrypt_room_message(&self, message: &str, room_key: &str) -> Result<String> {
        let key = secretbox::Key::from_slice(
            &base64::decode(room_key).map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid room key".to_string()))?;

        let nonce = secretbox::gen_nonce();
        let ciphertext = secretbox::seal(message.as_bytes(), &nonce, &key);

        let mut combined = nonce.as_ref().to_vec();
        combined.extend_from_slice(&ciphertext);

        Ok(base64::encode(&combined))
    }

    /// Decrypt room message (symmetric)
    pub fn decrypt_room_message(&self, encrypted_message: &str, room_key: &str) -> Result<String> {
        let combined =
            base64::decode(encrypted_message).map_err(|e| AppError::Encryption(e.to_string()))?;

        if combined.len() < secretbox::NONCEBYTES {
            return Err(AppError::Encryption(
                "Invalid encrypted message".to_string(),
            ));
        }

        let nonce = secretbox::Nonce::from_slice(&combined[..secretbox::NONCEBYTES])
            .ok_or_else(|| AppError::Encryption("Invalid nonce".to_string()))?;

        let ciphertext = &combined[secretbox::NONCEBYTES..];

        let key = secretbox::Key::from_slice(
            &base64::decode(room_key).map_err(|e| AppError::Encryption(e.to_string()))?,
        )
        .ok_or_else(|| AppError::Encryption("Invalid room key".to_string()))?;

        let decrypted = secretbox::open(ciphertext, &nonce, &key)
            .map_err(|_| AppError::Encryption("Decryption failed".to_string()))?;

        String::from_utf8(decrypted).map_err(|e| AppError::Encryption(e.to_string()))
    }

    /// Generate random room key
    pub fn generate_room_key(&self) -> String {
        let key = secretbox::gen_key();
        base64::encode(key.as_ref())
    }

    /// Hash data
    pub fn hash(&self, data: &str) -> String {
        use sodiumoxide::crypto::generichash;
        let hash = generichash::hash(data.as_bytes(), None, None).unwrap();
        base64::encode(hash.as_ref())
    }

    /// Generate random bytes
    pub fn random_bytes(&self, length: usize) -> String {
        let bytes = randombytes::randombytes(length);
        base64::encode(&bytes)
    }
}
